source/common/grpc/common.cc

```diff
@@ -269,5 +286,98 @@ std::string Common::typeUrl(const std::string& qualified_name) {
   return typeUrlPrefix() + "/" + qualified_name;
 }
 
+struct BufferInstanceContainer {
+  BufferInstanceContainer(int ref_count, Buffer::InstancePtr buffer)
+      : ref_count_(ref_count), buffer_(std::move(buffer)) {}
+  std::atomic<int> ref_count_;
+  Buffer::InstancePtr buffer_;
+};
+
+static void derefBufferInstanceContainer(void* container_ptr) {
+  auto container = reinterpret_cast<BufferInstanceContainer*>(container_ptr);
+  container->ref_count_--;
+  if (container->ref_count_ <= 0) {
+    delete container;
+  }
+}
+
+grpc::ByteBuffer Common::makeByteBuffer(Buffer::InstancePtr bufferInstance) {
+  if (!bufferInstance) {
+    return {};
+  }
+  Buffer::RawSlice oneRawSlice;
+  // NB: we need to pass in >= 1 in order to get the real "n" (see Buffer::Instance for details).
+  int nSlices = bufferInstance->getRawSlices(&oneRawSlice, 1);
+  if (nSlices <= 0) {
+    return {};
+  }
+  auto container = new BufferInstanceContainer{nSlices, std::move(bufferInstance)};
+  if (nSlices == 1) {
+    grpc::Slice oneSlice(oneRawSlice.mem_, oneRawSlice.len_, &derefBufferInstanceContainer,
+                         container);
+    return {&oneSlice, 1};
+  }
+  STACK_ARRAY(manyRawSlices, Buffer::RawSlice, nSlices);
+  container->buffer_->getRawSlices(manyRawSlices.begin(), nSlices);
+  std::vector<grpc::Slice> slices;
+  slices.reserve(nSlices);
+  for (int i = 0; i < nSlices; i++) {
+    slices.emplace_back(manyRawSlices[i].mem_, manyRawSlices[i].len_, &derefBufferInstanceContainer,
+                        container);
+  }
+  return {&slices[0], slices.size()};
+}
+
+struct ByteBufferContainer {
+  ByteBufferContainer(int ref_count) : ref_count_(ref_count) {}
+  ~ByteBufferContainer() { ::free(fragments); }
+  std::atomic<int> ref_count_;
+  Buffer::BufferFragmentImpl* fragments = nullptr;
+  std::vector<grpc::Slice> slices_;
+};
+
+Buffer::InstancePtr Common::makeBufferInstance(const grpc::ByteBuffer& byteBuffer) {
+  auto buffer = std::make_unique<Buffer::OwnedImpl>();
+  if (byteBuffer.Length() == 0) {
+    return buffer;
+  }
+  // NB: ByteBuffer::Dump moves the data out of the ByteBuffer so we need to ensure that the
+  // lifetime of the Slice(s) exceeds our Buffer::Instance.
+  std::vector<grpc::Slice> slices;
+  byteBuffer.Dump(&slices);
+  if (slices.size() == 0) {
+    return buffer;
+  }
+  auto container = new ByteBufferContainer(static_cast<int>(slices.size()));
```

Comment:
Nit: `auto* container`

Reply:
Begs the question why  use auto if you are going to add redundant decorators?  However if this is the style I will comply.

Done.